content/tutorials/spatial_transform_crs_2/index.Rmd

```diff
@@ -0,0 +1,495 @@
+```
+
+A specific area shown with `ggplot2` in `EPSG:31370`:
+
+```{r ggplot_water_points}
+ggplot() + 
+  geom_sf(data = watercourses, fill = "lightblue", colour = "lightblue") +
+  geom_sf(data = points_31370, fill = "red", shape = 21) +
+  lims(x = c(154e3, 155e3), y = c(218e3, 219e3)) +
+  coord_sf(datum = "EPSG:31370")
+```
+
+### Enforcing a specific pipeline
+
+We can request available transformation pipelines with `sf_proj_pipelines()`:
+
+```{r}
+pipelines <- sf_proj_pipelines("EPSG:28992", "EPSG:31370")
+```
+
+One can further limit this by specifying the area of interest (`aoi` argument), which we didn't do here.
+
+It is a dataframe with following structure:
+
+```{r}
+glimpse(pipelines)
+```
+
+Some metadata about the pipelines:
+
+```{r paged.print=FALSE}
+pipelines %>% 
+  as_tibble %>% 
+  select(definition, description, grid_count, accuracy, instantiable)
+```
+
+We can see that the first one has best accuracy.
+Several pipelines are based on one or two grids for a better accuracy.
+
+When printing the `pipelines` object, its first row is printed in a more readable manner:
+
+```{r}
+pipelines
+```
+
+Since the rows of pipelines are sorted by `instantiable` and `accuracy`, it makes sense to designate the printed pipeline as _Best instantiable operation_.
+
+So the same information appears as when manually selecting and then printing, except for the number of 'candidate operations found':
+
+```{r}
+pipelines[1,]
+```
+
+If you'd like to get even more metadata on the available pipelines, including the WKT definitions, you can also run PROJ's [`projinfo`](https://proj.org/apps/projinfo.html) command line program from within R if it's available in your system `PATH`.
+
+```{r eval=FALSE}
+system("projinfo -s EPSG:28992 -t EPSG:31370 --spatial-test intersects -o WKT2:2019")
+```
+
+The *default* operation returned by PROJ's `projinfo`, based on the union of the source and target CRS's bounding box, appears to be just one pipeline with so-called [ballpark accuracy](https://proj.org/glossary.html#term-Ballpark-transformation).
+It's obtained by dropping the `--spatial-test intersects` option (or replacing `intersects` by `contains`, which is default), and so it matches the pipeline where `accuracy` equals `NA` in our `pipelines` dataframe:
+
+```{r}
+pipelines[13,]
+```
+
+The low accuracy of a ballpark transformation is because datum differences between source and target CRSs are neglected:
+
+> It does not attempt any datum shift, hence the "ballpark" qualifier in its name.
+> Its accuracy is unknown, and could lead in some cases to errors of a few hundreds of metres.
+
+*(Source: <https://proj.org/glossary.html#term-Ballpark-transformation>)*
+
+With `st_transform()` we can enforce a specific pipeline.
+Note that it is only available for `sfc` objects, not `sf` objects.
+
+Let's apply the most accurate pipeline to *all* points:
+
+```{r}
+chosen_pipeline_definition <- pipelines %>% filter(accuracy < 0.02) %>% pull(definition)
```

Comment:
maybe use `slice(1)` instead of `filter()` to make it very clear you chose the first row.

Reply:
Well, I'm going to disappoint :wink:. During experimentation with installing grids, removing grids, turning on and off PROJ's networking capability, this sorting turned out to be rather unstable, at least with PROJ 7.2.1, because sorting could e.g. still be according to an older state.

Example: without having the grids for the most accurate transformation (0.011 m), the one with 0.011 was on line 9, which is where the non-instantiable pipelines start. That's OK, but this sorting was not (always) switched within the same R session after installing grids; the instantiable field _was_ updated to `TRUE` in that case, nevertheless! That is no doubt some buggy behaviour, but at the same time it's undoubtedly coming from PROJ, so investigating and posting such issue should apply to current PROJ (8.1.1), meaning that I should recompile sf with PROJ 8.1.1, re-investigate etc. - which is all doable but my time is not endless :wink:. So filtering on attribute always worked :smile_cat:.

Reply:
I'l add a comment in the R chunk that explains this is indeed the line with 0.011 m accuracy, which should be the first line if grids are installed.

Reply:
Done in 8e9eabbb